<a href="https://colab.research.google.com/github/farhan1503001/Breast-Cancer-Classification-From-Histopathological-Images/blob/master/BreakHTS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import tensorflow as tf
import h5py
import numpy as np
import os
import cv2
from sklearn.metrics import precision_score,recall_score,f1_score,roc_auc_score
from imblearn.metrics import sensitivity_specificity_support
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()

/usr/local/lib/python3.6/dist-packages/sklearn/externals/six.py:31: FutureWarning: The module is deprecated in version 0.21 and will be removed in version 0.23 since we've dropped support for Python 2.7. Please rely on the official version of six (https://pypi.org/project/six/).
  "(https://pypi.org/project/six/).", FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.neighbors.base module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.neighbors. Anything that cannot be imported from sklearn.neighbors is now part of the private API.
  warnings.warn(message, FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.metrics.classification module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be import

In [ ]:
x_dataset=[]
y_dataset=[]
src=os.listdir('/content/drive/My Drive/40/benign/')
i=0
for filename in src:
    full_file_name='/content/drive/My Drive/40/benign/'+filename
    x=cv2.imread(full_file_name)
    x=np.array(x,dtype='float32')
    y=0
    y=np.array(y,dtype='float32')
    x=x/255.0
    x=cv2.resize(x,(224,224))
    x_dataset.append(x)
    y_dataset.append(y)
    i=i+1
    print(i)
    

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277


In [ ]:
src=os.listdir('/content/drive/My Drive/40/malignant/')
i=0
for filename in src:
    full_file_name='/content/drive/My Drive/40/malignant/'+filename
    x=cv2.imread(full_file_name)
    x=np.array(x,dtype='float32')
    y=1
    y=np.array(y,dtype='float32')
    x=x/255.0
    x=cv2.resize(x,(224,224))
    x_dataset.append(x)
    y_dataset.append(y)
    i=i+1
    print(i)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277


In [ ]:
X_dataset=np.array(x_dataset)
Y_dataset=np.array(y_dataset)
Y_dataset=Y_dataset.reshape((2013,1))

In [ ]:
X_dataset[1]

array([[[0.88267463, 0.8660102 , 0.8881259 ],
        [0.8500372 , 0.8293155 , 0.8530156 ],
        [0.84982276, 0.84957767, 0.8556154 ],
        ...,
        [0.786036  , 0.6952294 , 0.84797573],
        [0.74738276, 0.6742013 , 0.8188726 ],
        [0.7414216 , 0.65931374, 0.7855392 ]],

       [[0.7914325 , 0.7414676 , 0.7829591 ],
        [0.75962013, 0.728543  , 0.76251316],
        [0.7755821 , 0.76143867, 0.7828082 ],
        ...,
        [0.8450346 , 0.7365131 , 0.87455136],
        [0.7745799 , 0.6999935 , 0.831169  ],
        [0.69990593, 0.6182488 , 0.74552697]],

       [[0.61046046, 0.56021976, 0.60584295],
        [0.69322264, 0.6308627 , 0.6789719 ],
        [0.8314667 , 0.7980218 , 0.84020704],
        ...,
        [0.82397807, 0.71914613, 0.8603226 ],
        [0.79241514, 0.6968269 , 0.83481705],
        [0.73150384, 0.6453191 , 0.75698316]],

       ...,

       [[0.87794363, 0.89410913, 0.8987659 ],
        [0.83353925, 0.78013194, 0.8043245 ],
        [0.83500814, 0

In [ ]:
from sklearn.model_selection import KFold
folds=list(KFold(n_splits=5,shuffle=True,random_state=1).split(X_dataset,Y_dataset))

In [ ]:
Inception=tf.keras.applications.InceptionV3(include_top=False,input_shape=(224,224,3))
input_image=tf.keras.layers.Input((224,224,3))
x=Inception (input_image)
x=tf.keras.layers.Flatten()(x)
x=tf.keras.layers.Dense(2)(x)
out=tf.keras.layers.Activation(activation='softmax')(x)

model=tf.keras.Model(inputs=input_image,outputs=out)

model.compile(optimizer=tf.keras.optimizers.RMSprop(lr=0.0001),loss='sparse_categorical_crossentropy',metrics=['accuracy'])


87916544/87910968 [==============================] - 5s 0us/step


In [ ]:
from sklearn.preprocessing import label_binarize
from sklearn.metrics import auc,roc_curve

In [ ]:
History=[]
Prescore=[]
Recall=[]
F1=[]
Sensspeci=[]
Roc_auc=[]
for j,(train_idx,val_idx) in enumerate(folds):
    print("Fold "+str(j+1))
    
    x_train=X_dataset[train_idx]
    y_train=Y_dataset[train_idx]
    x_val=X_dataset[val_idx]
    y_val=Y_dataset[val_idx]
    history=model.fit(x_train,y_train,batch_size=32,epochs=20,validation_data=(x_val,y_val))
    y_predict=model.predict(x_val)
    Y_predict=np.argmax(y_predict,axis=1)
    print(y_val)
    l=len(y_val)
    Y_predict=np.reshape(Y_predict,(l,1))
    #print(Y_predict)
    prescore=precision_score(y_val,Y_predict)
    recaller=recall_score(y_val,Y_predict)
    score=f1_score(y_val,Y_predict)
    sensitivity=sensitivity_specificity_support(y_val,Y_predict,average='binary')
    print("Precision: ",prescore)
    print("Recall: ",recaller)
    print("F1-score ",score)
    print("Sensitivity ",sensitivity)
    History.append(history)
    Prescore.append(prescore)
    F1.append(score)
    Recall.append(recaller)
    Sensspeci.append(sensitivity)

Fold 1
Epoch 1/20
51/51 [==============================] - 37s 720ms/step - loss: 0.1956 - accuracy: 0.9280 - val_loss: 3.4018 - val_accuracy: 0.6873
Epoch 2/20
51/51 [==============================] - 32s 630ms/step - loss: 0.0419 - accuracy: 0.9851 - val_loss: 1.3923 - val_accuracy: 0.7519
Epoch 3/20
51/51 [==============================] - 32s 631ms/step - loss: 0.0260 - accuracy: 0.9907 - val_loss: 1.2239 - val_accuracy: 0.7965
Epoch 4/20
51/51 [==============================] - 32s 630ms/step - loss: 0.0192 - accuracy: 0.9919 - val_loss: 0.8790 - val_accuracy: 0.8586
Epoch 5/20
51/51 [==============================] - 32s 629ms/step - loss: 0.0286 - accuracy: 0.9913 - val_loss: 0.3251 - val_accuracy: 0.9355
Epoch 6/20
51/51 [==============================] - 32s 628ms/step - loss: 0.0358 - accuracy: 0.9913 - val_loss: 0.3375 - val_accuracy: 0.9380
Epoch 7/20
51/51 [==============================] - 32s 630ms/step - loss: 0.0169 - accuracy: 0.9938 - val_loss: 0.3197 - val_accuracy:

In [ ]:
History[0].history['val_accuracy']


[0.729528546333313,
 0.6848635077476501,
 0.6848635077476501,
 0.7692307829856873,
 0.7940446734428406,
 0.9007444381713867,
 0.9503722190856934,
 0.92555832862854,
 0.9181141257286072,
 0.9478908181190491]

In [ ]:
val_acc=History[0].history['val_accuracy']
val_acc1=History[1].history['val_accuracy']
val_acc2=History[2].history['val_accuracy']
val_acc3=History[3].history['val_accuracy']
val_acc4=History[4].history['val_accuracy']
Accuracvalid1=val_acc[19]
Accuracyvalid2=val_acc1[19]
Accuracyvalid3=val_acc2[19]
Accuracyvalid4=val_acc3[19]
Accuracyvalid5=val_acc4[19]
Mean_valid_Accuracy=(Accuracvalid1+Accuracyvalid2+Accuracyvalid3+Accuracyvalid4+Accuracyvalid5)/5.00
print("validation Accuracy for Fold-1 ",Accuracvalid1)
print("validation Accuracy for Fold-2",Accuracyvalid2)
print("Validation Accuracy for Fold-3 ",Accuracyvalid3)
print("Validation Accuracy for Fold-4  ",Accuracyvalid4)
print("Validation Accuracy for Fold-5  ",Accuracyvalid5)
print("Mean Validation Accuracy ",Mean_valid_Accuracy)

validation Accuracy for Fold-1  0.9727047085762024
validation Accuracy for Fold-2 0.9975185990333557
Validation Accuracy for Fold-3  1.0
Validation Accuracy for Fold-4   1.0
Validation Accuracy for Fold-5   1.0
Mean Validation Accuracy  0.9940446615219116


# Accuracy in run 1
validation Accuracy for Fold-1  0.9478908181190491

validation Accuracy for Fold-2 0.9950371980667114

Validation Accuracy for Fold-3  0.9975185990333557

Validation Accuracy for Fold-4   0.9950248599052429

Validation Accuracy for Fold-5   0.9950248599052429

Mean Validation Accuracy  0.9860992670059204

In [ ]:
print("Printing the precision scores and average of them")
avg=0.0
for i in range(5):
  print("Precision of Fold -",i+1,Prescore[i])
  avg=avg+Prescore[i]

print("Average precision-score ",avg/5.00)


Printing the precision scores and average of them
Precision of Fold - 1 0.974910394265233
Precision of Fold - 2 1.0
Precision of Fold - 3 1.0
Precision of Fold - 4 1.0
Precision of Fold - 5 1.0
Average precision-score  0.9949820788530467


In [ ]:
print("Printing the recall scores and average of them")
avgr=0.0
for i in range(5):
  print("Precision of Fold -",i+1,Recall[i])
  avgr=avgr+Recall[i]

print("Average precision-score ",avgr/5.00)


Printing the recall scores and average of them
Precision of Fold - 1 0.9855072463768116
Precision of Fold - 2 0.996415770609319
Precision of Fold - 3 1.0
Precision of Fold - 4 1.0
Precision of Fold - 5 1.0
Average precision-score  0.9963846033972261


In [ ]:
print("Printing the F-1 scores and average of them")
avgf=0.0
for i in range(5):
  print("Precision of Fold -",i+1,F1[i])
  avgf=avgf+F1[i]

print("Average f1-score ",avgf/5.00)

Printing the F-1 scores and average of them
Precision of Fold - 1 0.9801801801801803
Precision of Fold - 2 0.9982046678635548
Precision of Fold - 3 1.0
Precision of Fold - 4 1.0
Precision of Fold - 5 1.0
Average f1-score  0.995676969608747


In [ ]:
print("Printing the sensitivity scores and average of them")
avgsen=0.0
for i in range(5):
  print("Precision of Fold -",i+1,Sensspeci[i][0])
  avgsen=avgsen+Sensspeci[i][0]

print("Average sensitivity-score ",avgsen/5.00)

Printing the sensitivity scores and average of them
Precision of Fold - 1 0.9855072463768116
Precision of Fold - 2 0.996415770609319
Precision of Fold - 3 1.0
Precision of Fold - 4 1.0
Precision of Fold - 5 1.0
Average sensitivity-score  0.9963846033972261


In [ ]:
print("Printing the specificity scores and average of them")
avgsp=0.0
for i in range(5):
  print("Precision of Fold -",i+1,Sensspeci[i][1])
  avgsp=avgsp+Sensspeci[i][1]

print("Average specificity-score ",avgsp/5.00)

Printing the specificity scores and average of them
Precision of Fold - 1 0.9448818897637795
Precision of Fold - 2 1.0
Precision of Fold - 3 1.0
Precision of Fold - 4 1.0
Precision of Fold - 5 1.0
Average specificity-score  0.988976377952756


In [ ]:
y_predict=model.predict(X_dataset)

In [ ]:
Y_predict=np.argmax(y_predict,axis=1)

In [ ]:
Y_predict=np.reshape(Y_predict,(2013,1))
print(Y_predict)

[[0]
 [0]
 [0]
 ...
 [1]
 [1]
 [1]]


In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
metrics=confusion_matrix(Y_dataset,Y_predict)
print(metrics)

[[ 625    0]
 [   0 1388]]
